File: Doit_020.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;

public class Doit_020 {

    private static int[] arr;
    private static int[] tmp;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        int N = Integer.parseInt(br.readLine());

        arr = new int[N];
        tmp = new int[N];
        for (int idx = 0; idx < N; idx++) {
            arr[idx] = Integer.parseInt(br.readLine());
        }

        mergeSort(0, N - 1);
        for (int idx = 0; idx < N; idx++) {
            bw.write(String.valueOf(arr[idx]));
            bw.newLine();
        }
        bw.flush();
        bw.close();
    }
    private static void mergeSort(int start, int end) {
        if (start < end) {
            int mid = (start + end) / 2;
            mergeSort(start, mid);
            mergeSort(mid + 1, end);
            merge(start, mid, end);
        }
    }
    private static void merge(int start, int mid, int end) {
        int index = start;
        int pointer1 = start;
        int pointer2 = mid + 1;
        while (pointer1 <= mid && pointer2 <= end) {
            if (arr[pointer1] <= arr[pointer2]) {
                tmp[index] = arr[pointer1];
                pointer1++;
            } else {
                tmp[index] = arr[pointer2];
                pointer2++;
            }
            index++;
        }
        while (pointer1 <= mid) {
            tmp[index] = arr[pointer1];
            index++;
            pointer1++;
        }
        while (pointer2 <= end) {
            tmp[index] = arr[pointer2];
            index++;
            pointer2++;
        }
        for (int i = start; i <= end; i++) {
            arr[i] = tmp[i];
        }
    }
}

File: Doit_020.cpp
#include <iostream>
#include <vector>

using namespace std;
// 병합 정렬 -> O(nlog)

static vector<int> vec;
static vector<int> tmp;

void merge_sort(int start, int end) {
  if (end - start < 1) return;

  int mid = (start + end) / 2;
  merge_sort(start, mid);
  merge_sort(mid + 1, end);

  for (int i = start; i <= end; i++) tmp[i] = vec[i];  // 복사

  int k = start;
  int i = start;
  int j = mid + 1;

  while (i <= mid && j <= end) {  // 두 그룹 병합 로직 -> 정렬
    if (tmp[i] <= tmp[j]) {
      vec[k++] = tmp[i++];
    } else {
      vec[k++] = tmp[j++];
    }
  }

  while (i <= mid) vec[k++] = tmp[i++];
  while (j <= end) vec[k++] = tmp[j++];
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);

  int n;
  cin >> n;

  vec = vector<int>(n + 1, 0);
  tmp = vector<int>(n + 1, 0);

  for (int i = 1; i <= n; i++) {
    cin >> vec[i];
  }

  merge_sort(1, n);  // 병합 정렬

  for (int i = 1; i <= n; i++) {
    cout << vec[i] << "\n";
  }
}
